File: python/artificer_client/resources/workflows.py
"""
Workflows resource
"""

from typing import Dict, Any, List, Optional
from .base import BaseResource


class Workflows(BaseResource):
    """Workflows API resource for Prefect orchestration."""

    def list(self) -> dict:
        """
        List all available Prefect workflows.

        Returns:
            Available workflows and Prefect service status

        Example:
            >>> workflows = client.workflows.list()
            >>> print(f"Available: {workflows['available']}")
            >>> for wf in workflows['workflows']:
            ...     print(f"{wf['id']}: {wf['description']}")
        """
        return self._trpc_request("workflows.list")

    def get(self, workflow_id: str) -> dict:
        """
        Get details for a specific workflow.

        Args:
            workflow_id: Workflow ID (e.g., 'pdf-to-html')

        Returns:
            Workflow details including inputs and description

        Example:
            >>> workflow = client.workflows.get("pdf-to-html")
            >>> print(f"Name: {workflow['name']}")
            >>> print(f"Inputs: {workflow['inputs']}")
        """
        return self._trpc_request("workflows.get", {"workflowId": workflow_id})

    def execute(
        self,
        workflow_id: str,
        inputs: Dict[str, Any]
    ) -> dict:
        """
        Execute any workflow by ID with custom inputs.

        Args:
            workflow_id: Workflow ID
            inputs: Workflow-specific inputs (see workflow.inputs for schema)

        Returns:
            Workflow execution result

        Example:
            >>> result = client.workflows.execute(
            ...     "pdf-to-html",
            ...     {
            ...         "pdf_data": base64_encoded_pdf,
            ...         "include_styles": True,
            ...         "title": "My Document"
            ...     }
            ... )
            >>> print(f"Success: {result['success']}")
            >>> print(f"Result: {result['result']}")
        """
        return self._trpc_request("workflows.execute", {
            "workflowId": workflow_id,
            "inputs": inputs
        })

    def execute_pdf_to_html(
        self,
        pdf_data: str,
        include_styles: bool = True,
        title: Optional[str] = None
    ) -> dict:
        """
        Execute PDF to HTML conversion pipeline.

        Workflow: Extract PDF → Convert to Portable Text → Export as HTML

        Args:
            pdf_data: Base64-encoded PDF file
            include_styles: Include CSS styles in HTML (default: True)
            title: Optional title for the HTML document

        Returns:
            Workflow result with HTML output

        Example:
            >>> import base64
            >>> with open("document.pdf", "rb") as f:
            ...     pdf_data = base64.b64encode(f.read()).decode()
            >>> result = client.workflows.execute_pdf_to_html(
            ...     pdf_data,
            ...     include_styles=True,
            ...     title="My Document"
            ... )
            >>> html = result['result']['html']
        """
        inputs = {
            "pdf_data": pdf_data,
            "include_styles": include_styles
        }
        if title:
            inputs["title"] = title

        return self._trpc_request("workflows.executePdfToHtml", inputs)

    def execute_pdf_with_ocr(
        self,
        pdf_data: str,
        chunk_size: int = 1000,
        chunk_overlap: int = 200
    ) -> dict:
        """
        Execute PDF processing with OCR and chunking.

        Workflow: Extract PDF → OCR if needed → Chunk text → Process

        Args:
            pdf_data: Base64-encoded PDF file
            chunk_size: Size of text chunks (default: 1000 characters)
            chunk_overlap: Overlap between chunks (default: 200 characters)

        Returns:
            Workflow result with chunks and metadata

        Example:
            >>> result = client.workflows.execute_pdf_with_ocr(
            ...     pdf_data,
            ...     chunk_size=500,
            ...     chunk_overlap=100
            ... )
            >>> chunks = result['result']['chunks']
            >>> print(f"Extracted {len(chunks)} chunks")
        """
        return self._trpc_request("workflows.executePdfWithOcr", {
            "pdf_data": pdf_data,
            "chunk_size": chunk_size,
            "chunk_overlap": chunk_overlap
        })

    def execute_batch_pdf(
        self,
        pdf_files: List[Dict[str, str]],
        max_workers: int = 3
    ) -> dict:
        """
        Execute batch PDF processing in parallel.

        Workflow: Process multiple PDFs concurrently with progress tracking

        Args:
            pdf_files: List of PDFs, each with 'filename' and 'data' (base64)
            max_workers: Maximum parallel workers (default: 3)

        Returns:
            Workflow result with all processed PDFs

        Example:
            >>> pdf_files = [
            ...     {"filename": "doc1.pdf", "data": base64_pdf1},
            ...     {"filename": "doc2.pdf", "data": base64_pdf2}
            ... ]
            >>> result = client.workflows.execute_batch_pdf(
            ...     pdf_files,
            ...     max_workers=5
            ... )
            >>> for item in result['result']['results']:
            ...     print(f"{item['filename']}: {item['status']}")
        """
        return self._trpc_request("workflows.executeBatchPdf", {
            "pdf_files": pdf_files,
            "max_workers": max_workers
        })

    def execute_image_ocr(
        self,
        image_data: str,
        language: str = "eng"
    ) -> dict:
        """
        Execute OCR on an image.

        Workflow: Preprocess image → Run OCR → Extract structured text

        Args:
            image_data: Base64-encoded image file
            language: OCR language code (default: 'eng')

        Returns:
            Workflow result with extracted text and confidence

        Example:
            >>> import base64
            >>> with open("receipt.jpg", "rb") as f:
            ...     image_data = base64.b64encode(f.read()).decode()
            >>> result = client.workflows.execute_image_ocr(
            ...     image_data,
            ...     language="eng"
            ... )
            >>> print(f"Text: {result['result']['text']}")
            >>> print(f"Confidence: {result['result']['confidence']}")
        """
        return self._trpc_request("workflows.executeImageOcr", {
            "image_data": image_data,
            "language": language
        })

    def execute_markdown_conversion(
        self,
        source_format: str,
        content: str,
        target_format: str = "markdown"
    ) -> dict:
        """
        Execute format conversion to/from markdown.

        Workflow: Parse source → Convert to AST → Export to target format

        Args:
            source_format: Source format ('html', 'docx', 'pdf', 'markdown')
            content: Content to convert
            target_format: Target format (default: 'markdown')

        Returns:
            Workflow result with converted content

        Example:
            >>> result = client.workflows.execute_markdown_conversion(
            ...     source_format="html",
            ...     content="<h1>Title</h1><p>Content</p>",
            ...     target_format="markdown"
            ... )
            >>> markdown = result['result']['content']
        """
        return self._trpc_request("workflows.executeMarkdownConversion", {
            "source_format": source_format,
            "content": content,
            "target_format": target_format
        })

    def execute_translation(
        self,
        text: str,
        target_language: str,
        source_language: str = "auto",
        use_specialists: bool = True,
        min_specialists_required: int = 3
    ) -> dict:
        """
        Execute translation with multiple specialist models.

        Workflow: Parallel translation by specialists → Consensus selection

        Args:
            text: Text to translate
            target_language: Target language code (e.g., 'es', 'fr', 'de')
            source_language: Source language (default: 'auto' for auto-detect)
            use_specialists: Use multiple models for consensus (default: True)
            min_specialists_required: Min successful translations (default: 3)

        Returns:
            Workflow result with translation and metadata

        Example:
            >>> result = client.workflows.execute_translation(
            ...     text="Hello, world!",
            ...     target_language="es",
            ...     use_specialists=True
            ... )
            >>> translation = result['result']['translation']
            >>> print(f"Confidence: {result['result']['confidence']}")
            >>> print(f"Models used: {result['result']['models_used']}")
        """
        return self._trpc_request("workflows.executeTranslation", {
            "text": text,
            "target_language": target_language,
            "source_language": source_language,
            "use_specialists": use_specialists,
            "min_specialists_required": min_specialists_required
        })

    def execute_batch_translation(
        self,
        texts: List[str],
        target_language: str,
        source_language: str = "auto",
        max_workers: int = 3
    ) -> dict:
        """
        Execute batch translation in parallel.

        Workflow: Process multiple texts concurrently with progress tracking

        Args:
            texts: List of texts to translate
            target_language: Target language code
            source_language: Source language (default: 'auto')
            max_workers: Maximum parallel workers (default: 3)

        Returns:
            Workflow result with all translations

        Example:
            >>> texts = [
            ...     "Hello, world!",
            ...     "How are you?",
            ...     "Good morning"
            ... ]
            >>> result = client.workflows.execute_batch_translation(
            ...     texts,
            ...     target_language="es",
            ...     max_workers=5
            ... )
            >>> for i, item in enumerate(result['result']['translations']):
            ...     print(f"{texts[i]} → {item['translation']}")
        """
        return self._trpc_request("workflows.executeBatchTranslation", {
            "texts": texts,
            "target_language": target_language,
            "source_language": source_language,
            "max_workers": max_workers
        })

    def health_check(self) -> dict:
        """
        Check Prefect service availability.

        Returns:
            Health check result

        Example:
            >>> health = client.workflows.health_check()
            >>> print(f"Available: {health['available']}")
            >>> print(f"Version: {health.get('version', 'unknown')}")
        """
        return self._trpc_request("workflows.healthCheck")

    # ========================================
    # CUSTOM WORKFLOWS - Phase 2
    # ========================================

    def register_custom_workflow(
        self,
        workflow_id: str,
        definition: Dict[str, Any]
    ) -> dict:
        """
        Register a custom workflow definition.

        Allows defining custom DAG workflows using declarative configuration
        that gets translated into executable Prefect flows.

        Args:
            workflow_id: Unique workflow identifier
            definition: Workflow definition dictionary with:
                - name: Workflow name
                - description: Optional description
                - version: Optional version (semver)
                - tasks: List of task definitions
                - output: Output mapping
                - options: Execution options (parallel, timeout, etc.)

        Returns:
            Registration result

        Example:
            >>> workflow_def = {
            ...     "name": "pdf-extract-and-chunk",
            ...     "description": "Extract PDF and chunk into segments",
            ...     "version": "1.0.0",
            ...     "tasks": [
            ...         {
            ...             "id": "extract",
            ...             "type": "extract_pdf_text",
            ...             "inputs": {"pdf_data": "{{workflow.input.pdf_data}}"}
            ...         },
            ...         {
            ...             "id": "chunk",
            ...             "type": "chunk_document",
            ...             "depends_on": ["extract"],
            ...             "inputs": {
            ...                 "document_id": "{{workflow.input.document_id}}",
            ...                 "project_id": "{{workflow.input.project_id}}",
            ...                 "content": "{{extract.text}}",
            ...                 "chunk_size": 1000,
            ...                 "chunk_overlap": 200
            ...             }
            ...         }
            ...     ],
            ...     "output": {
            ...         "chunks": "{{chunk.chunks}}",
            ...         "metadata": "{{extract.metadata}}"
            ...     }
            ... }
            >>> result = client.workflows.register_custom_workflow(
            ...     "my-pdf-workflow",
            ...     workflow_def
            ... )
            >>> print(f"Registered: {result['workflowId']}")
        """
        return self._trpc_request("workflows.registerCustomWorkflow", {
            "workflowId": workflow_id,
            "definition": definition
        })

    def list_custom_workflows(self) -> dict:
        """
        List all registered custom workflows.

        Returns:
            List of custom workflows with metadata

        Example:
            >>> workflows = client.workflows.list_custom_workflows()
            >>> for wf in workflows['workflows']:
            ...     print(f"{wf['id']}: {wf['name']} ({wf['taskCount']} tasks)")
        """
        return self._trpc_request("workflows.listCustomWorkflows")

    def get_custom_workflow(self, workflow_id: str) -> dict:
        """
        Get a custom workflow definition.

        Args:
            workflow_id: Workflow ID

        Returns:
            Workflow definition

        Example:
            >>> workflow = client.workflows.get_custom_workflow("my-pdf-workflow")
            >>> print(f"Name: {workflow['name']}")
            >>> print(f"Tasks: {len(workflow['tasks'])}")
        """
        return self._trpc_request("workflows.getCustomWorkflow", {
            "workflowId": workflow_id
        })

    def execute_custom_workflow(
        self,
        workflow_id: str,
        inputs: Dict[str, Any]
    ) -> dict:
        """
        Execute a registered custom workflow.

        Args:
            workflow_id: Workflow ID
            inputs: Workflow inputs (referenced as {{workflow.input.key}})

        Returns:
            Workflow execution result

        Example:
            >>> result = client.workflows.execute_custom_workflow(
            ...     "my-pdf-workflow",
            ...     {
            ...         "pdf_data": base64_encoded_pdf,
            ...         "document_id": "doc123",
            ...         "project_id": "proj456"
            ...     }
            ... )
            >>> chunks = result['result']['chunks']
            >>> print(f"Processed into {len(chunks)} chunks")
        """
        return self._trpc_request("workflows.executeCustomWorkflow", {
            "workflowId": workflow_id,
            "inputs": inputs
        })

    def delete_custom_workflow(self, workflow_id: str) -> dict:
        """
        Delete a custom workflow.

        Args:
            workflow_id: Workflow ID

        Returns:
            Deletion result

        Example:
            >>> result = client.workflows.delete_custom_workflow("my-pdf-workflow")
            >>> print(result['message'])
        """
        return self._trpc_request("workflows.deleteCustomWorkflow", {
            "workflowId": workflow_id
        })

    def validate_workflow_definition(self, definition: Dict[str, Any]) -> dict:
        """
        Validate a workflow definition without registering it.

        Useful for checking if a workflow definition is valid before registration.

        Args:
            definition: Workflow definition dictionary

        Returns:
            Validation result with 'valid' boolean and optional 'error' message

        Example:
            >>> workflow_def = {
            ...     "name": "test-workflow",
            ...     "tasks": [
            ...         {
            ...             "id": "task1",
            ...             "type": "extract_pdf_text",
            ...             "inputs": {"pdf_data": "{{workflow.input.pdf}}"}
            ...         }
            ...     ]
            ... }
            >>> result = client.workflows.validate_workflow_definition(workflow_def)
            >>> if result['valid']:
            ...     print("✓ Workflow definition is valid")
            ... else:
            ...     print(f"✗ Validation failed: {result['error']}")
        """
        return self._trpc_request("workflows.validateWorkflowDefinition", {
            "definition": definition
        })

File: src/server/routers/workflows.ts
/**
 * Workflows Router - Prefect workflow orchestration
 *
 * Exposes pre-built Prefect workflows via API for:
 * - Document processing pipelines
 * - Translation workflows
 * - Batch operations
 * - Custom DAG execution
 */

import { z } from 'zod';
import { router, publicProcedure } from '../trpc';
import { PrefectService, WorkflowInput } from '../services/workflows/PrefectService';

const prefectService = new PrefectService();

export const workflowsRouter = router({
  /**
   * List all available workflows
   */
  list: publicProcedure.query(async () => {
    return {
      workflows: prefectService.listWorkflows(),
      available: await prefectService.isAvailable(),
    };
  }),

  /**
   * Get workflow details
   */
  get: publicProcedure
    .input(
      z.object({
        workflowId: z.string().describe('Workflow ID'),
      })
    )
    .query(({ input }) => {
      const workflow = prefectService.getWorkflow(input.workflowId);

      if (!workflow) {
        throw new Error(`Workflow not found: ${input.workflowId}`);
      }

      return {
        id: input.workflowId,
        ...workflow,
      };
    }),

  /**
   * Execute a workflow
   */
  execute: publicProcedure
    .input(
      z.object({
        workflowId: z.string().describe('Workflow ID to execute'),
        inputs: z.record(z.any()).describe('Workflow inputs'),
      })
    )
    .mutation(async ({ input }) => {
      const result = await prefectService.executeWorkflow(input.workflowId, input.inputs);

      if (!result.success) {
        throw new Error(result.error || 'Workflow execution failed');
      }

      return result;
    }),

  /**
   * Execute PDF to HTML pipeline
   */
  executePdfToHtml: publicProcedure
    .input(
      z.object({
        pdfData: z.string().describe('Base64 encoded PDF data'),
        includeStyles: z.boolean().optional().default(true),
        title: z.string().optional(),
      })
    )
    .mutation(async ({ input }) => {
      return prefectService.executeWorkflow('pdf-to-html', {
        pdf_data: input.pdfData,
        include_styles: input.includeStyles,
        title: input.title,
      });
    }),

  /**
   * Execute PDF with OCR pipeline
   */
  executePdfWithOcr: publicProcedure
    .input(
      z.object({
        pdfData: z.string().describe('Base64 encoded PDF data'),
        chunkSize: z.number().optional().default(1000),
        chunkOverlap: z.number().optional().default(200),
      })
    )
    .mutation(async ({ input }) => {
      return prefectService.executeWorkflow('pdf-with-ocr', {
        pdf_data: input.pdfData,
        chunk_size: input.chunkSize,
        chunk_overlap: input.chunkOverlap,
      });
    }),

  /**
   * Execute batch PDF processing
   */
  executeBatchPdf: publicProcedure
    .input(
      z.object({
        files: z
          .array(
            z.object({
              filename: z.string(),
              data: z.string().describe('Base64 encoded PDF data'),
            })
          )
          .min(1)
          .max(100),
      })
    )
    .mutation(async ({ input }) => {
      return prefectService.executeWorkflow('batch-pdf-processing', {
        pdf_files: input.files,
      });
    }),

  /**
   * Execute image OCR pipeline
   */
  executeImageOcr: publicProcedure
    .input(
      z.object({
        images: z
          .array(
            z.object({
              data: z.string().describe('Base64 encoded image data'),
              contentType: z.string().optional().default('image/png'),
            })
          )
          .min(1)
          .max(100),
        minConfidence: z.number().min(0).max(1).optional().default(0.7),
      })
    )
    .mutation(async ({ input }) => {
      return prefectService.executeWorkflow('image-ocr', {
        images: input.images.map((img) => ({
          data: img.data,
          content_type: img.contentType,
        })),
        min_confidence: input.minConfidence,
      });
    }),

  /**
   * Execute markdown conversion pipeline
   */
  executeMarkdownConversion: publicProcedure
    .input(
      z.object({
        markdownContent: z.string().min(1).max(1000000),
      })
    )
    .mutation(async ({ input }) => {
      return prefectService.executeWorkflow('markdown-conversion', {
        markdown_content: input.markdownContent,
      });
    }),

  /**
   * Execute translation pipeline
   */
  executeTranslation: publicProcedure
    .input(
      z.object({
        text: z.string().min(1).max(100000),
        language: z.enum(['kor', 'jpn', 'chi']),
        minSuccessfulSpecialists: z.number().min(1).max(5).optional().default(3),
        selectionStrategy: z
          .enum(['ensemble', 'vote', 'quality_score'])
          .optional()
          .default('ensemble'),
      })
    )
    .mutation(async ({ input }) => {
      return prefectService.executeWorkflow('translation', {
        text: input.text,
        language: input.language,
        min_successful_specialists: input.minSuccessfulSpecialists,
        selection_strategy: input.selectionStrategy,
      });
    }),

  /**
   * Execute batch translation
   */
  executeBatchTranslation: publicProcedure
    .input(
      z.object({
        documents: z
          .array(
            z.object({
              id: z.string(),
              text: z.string(),
            })
          )
          .min(1)
          .max(100),
        language: z.enum(['kor', 'jpn', 'chi']),
        minSuccessful: z.number().min(1).max(5).optional().default(3),
      })
    )
    .mutation(async ({ input }) => {
      return prefectService.executeWorkflow('batch-translation', {
        documents: input.documents,
        language: input.language,
        min_successful: input.minSuccessful,
      });
    }),

  /**
   * Check workflow system health
   */
  healthCheck: publicProcedure.query(async () => {
    const available = await prefectService.isAvailable();

    return {
      available,
      status: available ? 'healthy' : 'unavailable',
      message: available
        ? 'Prefect workflows are available'
        : 'Prefect is not installed or not accessible',
    };
  }),

  /**
   * CUSTOM WORKFLOWS - Phase 2: Declarative config → Prefect translation
   */

  /**
   * Register a custom workflow definition
   */
  registerCustomWorkflow: publicProcedure
    .input(
      z.object({
        workflowId: z.string().describe('Unique workflow ID'),
        definition: z.object({
          name: z.string(),
          description: z.string().optional(),
          version: z.string().optional(),
          tasks: z.array(
            z.object({
              id: z.string(),
              type: z.string(),
              inputs: z.record(z.any()),
              depends_on: z.array(z.string()).optional(),
              outputs: z.array(z.string()).optional(),
            })
          ),
          output: z.record(z.string()).optional(),
          options: z
            .object({
              parallel: z.boolean().optional(),
              retry_failed_tasks: z.boolean().optional(),
              timeout: z.number().optional(),
              task_runner: z.enum(['concurrent', 'sequential', 'dask']).optional(),
              max_retries: z.number().optional(),
            })
            .optional(),
          metadata: z.record(z.any()).optional(),
        }),
      })
    )
    .mutation(async ({ input }) => {
      // Validate workflow definition
      const validation = await prefectService.validateWorkflowDefinition(input.definition);

      if (!validation.valid) {
        throw new Error(`Invalid workflow definition: ${validation.error}`);
      }

      // Register workflow
      prefectService.registerCustomWorkflow(input.workflowId, input.definition);

      return {
        success: true,
        workflowId: input.workflowId,
        message: `Custom workflow ${input.definition.name} registered successfully`,
      };
    }),

  /**
   * List all custom workflows
   */
  listCustomWorkflows: publicProcedure.query(() => {
    return {
      workflows: prefectService.listCustomWorkflows(),
    };
  }),

  /**
   * Get a custom workflow definition
   */
  getCustomWorkflow: publicProcedure
    .input(
      z.object({
        workflowId: z.string().describe('Workflow ID'),
      })
    )
    .query(({ input }) => {
      const workflow = prefectService.getCustomWorkflow(input.workflowId);

      if (!workflow) {
        throw new Error(`Custom workflow not found: ${input.workflowId}`);
      }

      return {
        id: input.workflowId,
        ...workflow,
      };
    }),

  /**
   * Execute a custom workflow
   */
  executeCustomWorkflow: publicProcedure
    .input(
      z.object({
        workflowId: z.string().describe('Workflow ID'),
        inputs: z.record(z.any()).describe('Workflow inputs'),
      })
    )
    .mutation(async ({ input }) => {
      const result = await prefectService.executeCustomWorkflow(
        input.workflowId,
        input.inputs
      );

      if (!result.success) {
        throw new Error(result.error || 'Workflow execution failed');
      }

      return result;
    }),

  /**
   * Delete a custom workflow
   */
  deleteCustomWorkflow: publicProcedure
    .input(
      z.object({
        workflowId: z.string().describe('Workflow ID'),
      })
    )
    .mutation(({ input }) => {
      const deleted = prefectService.deleteCustomWorkflow(input.workflowId);

      if (!deleted) {
        throw new Error(`Custom workflow not found: ${input.workflowId}`);
      }

      return {
        success: true,
        workflowId: input.workflowId,
        message: 'Custom workflow deleted successfully',
      };
    }),

  /**
   * Validate a workflow definition without registering
   */
  validateWorkflowDefinition: publicProcedure
    .input(
      z.object({
        definition: z.object({
          name: z.string(),
          description: z.string().optional(),
          version: z.string().optional(),
          tasks: z.array(
            z.object({
              id: z.string(),
              type: z.string(),
              inputs: z.record(z.any()),
              depends_on: z.array(z.string()).optional(),
              outputs: z.array(z.string()).optional(),
            })
          ),
          output: z.record(z.string()).optional(),
          options: z.record(z.any()).optional(),
          metadata: z.record(z.any()).optional(),
        }),
      })
    )
    .mutation(async ({ input }) => {
      const validation = await prefectService.validateWorkflowDefinition(input.definition);

      return {
        valid: validation.valid,
        error: validation.error,
      };
    }),

  /**
   * WORKFLOW TEMPLATES - Phase 3: Pre-built workflow patterns
   */

  /**
   * List all workflow templates
   */
  listTemplates: publicProcedure
    .input(
      z
        .object({
          category: z.string().optional().describe('Filter by category'),
        })
        .optional()
    )
    .query(async ({ input }) => {
      const templates = await prefectService.listTemplates(input?.category);

      return {
        templates,
        categories: await prefectService.getTemplateCategories(),
      };
    }),

  /**
   * Get a workflow template
   */
  getTemplate: publicProcedure
    .input(
      z.object({
        templateId: z.string().describe('Template ID'),
      })
    )
    .query(async ({ input }) => {
      const template = await prefectService.getTemplate(input.templateId);

      if (!template) {
        throw new Error(`Template not found: ${input.templateId}`);
      }

      return template;
    }),

  /**
   * Get template categories
   */
  getTemplateCategories: publicProcedure.query(async () => {
    const categories = await prefectService.getTemplateCategories();

    return {
      categories,
    };
  }),

  /**
   * Instantiate a template with parameters
   */
  instantiateTemplate: publicProcedure
    .input(
      z.object({
        templateId: z.string().describe('Template ID'),
        params: z.record(z.any()).describe('Template parameters'),
        autoRegister: z.boolean().optional().describe('Automatically register the workflow'),
        workflowId: z.string().optional().describe('Workflow ID for auto-registration'),
      })
    )
    .mutation(async ({ input }) => {
      // Instantiate template
      const definition = await prefectService.instantiateTemplate(
        input.templateId,
        input.params
      );

      // Optionally auto-register
      if (input.autoRegister && input.workflowId) {
        // Validate
        const validation = await prefectService.validateWorkflowDefinition(definition);

        if (!validation.valid) {
          throw new Error(`Invalid workflow definition: ${validation.error}`);
        }

        // Register
        prefectService.registerCustomWorkflow(input.workflowId, definition);

        return {
          definition,
          registered: true,
          workflowId: input.workflowId,
        };
      }

      return {
        definition,
        registered: false,
      };
    }),

  /**
   * ASYNC EXECUTION - Phase 4: Hybrid execution model
   */

  /**
   * Execute workflow asynchronously (background)
   */
  executeAsync: publicProcedure
    .input(
      z.object({
        workflowId: z.string().describe('Workflow ID'),
        inputs: z.record(z.any()).describe('Workflow inputs'),
        webhook: z
          .object({
            url: z.string().url(),
            method: z.enum(['POST', 'PUT']).optional(),
            headers: z.record(z.string()).optional(),
          })
          .optional()
          .describe('Webhook for completion notification'),
        priority: z.enum(['low', 'normal', 'high']).optional(),
      })
    )
    .mutation(async ({ input }) => {
      const jobId = await prefectService.executeWorkflowAsync(input.workflowId, input.inputs, {
        webhook: input.webhook,
        priority: input.priority,
      });

      return {
        jobId,
        status: 'PENDING',
        message: 'Workflow queued for execution',
      };
    }),

  /**
   * Execute custom workflow asynchronously
   */
  executeCustomAsync: publicProcedure
    .input(
      z.object({
        workflowId: z.string().describe('Custom workflow ID'),
        inputs: z.record(z.any()).describe('Workflow inputs'),
        webhook: z
          .object({
            url: z.string().url(),
            method: z.enum(['POST', 'PUT']).optional(),
            headers: z.record(z.string()).optional(),
          })
          .optional(),
        priority: z.enum(['low', 'normal', 'high']).optional(),
      })
    )
    .mutation(async ({ input }) => {
      const jobId = await prefectService.executeCustomWorkflowAsync(
        input.workflowId,
        input.inputs,
        {
          webhook: input.webhook,
          priority: input.priority,
        }
      );

      return {
        jobId,
        status: 'PENDING',
        message: 'Custom workflow queued for execution',
      };
    }),

  /**
   * Get job status
   */
  getJobStatus: publicProcedure
    .input(
      z.object({
        jobId: z.string().describe('Job ID'),
      })
    )
    .query(({ input }) => {
      const job = prefectService.getJobStatus(input.jobId);

      if (!job) {
        throw new Error(`Job not found: ${input.jobId}`);
      }

      return job;
    }),

  /**
   * List jobs
   */
  listJobs: publicProcedure
    .input(
      z
        .object({
          status: z
            .enum(['PENDING', 'RUNNING', 'COMPLETED', 'FAILED', 'CANCELLED', 'TIMEOUT'])
            .optional(),
          workflowId: z.string().optional(),
          workflowType: z.enum(['pre-built', 'custom', 'template']).optional(),
          limit: z.number().optional(),
        })
        .optional()
    )
    .query(({ input }) => {
      const jobs = prefectService.listJobs(input);

      return {
        jobs,
        total: jobs.length,
      };
    }),

  /**
   * Cancel job
   */
  cancelJob: publicProcedure
    .input(
      z.object({
        jobId: z.string().describe('Job ID'),
      })
    )
    .mutation(({ input }) => {
      const cancelled = prefectService.cancelJob(input.jobId);

      if (!cancelled) {
        throw new Error(`Cannot cancel job: ${input.jobId}`);
      }

      return {
        success: true,
        jobId: input.jobId,
        message: 'Job cancelled successfully',
      };
    }),

  /**
   * Delete job
   */
  deleteJob: publicProcedure
    .input(
      z.object({
        jobId: z.string().describe('Job ID'),
      })
    )
    .mutation(({ input }) => {
      const deleted = prefectService.deleteJob(input.jobId);

      if (!deleted) {
        throw new Error(`Cannot delete job: ${input.jobId}`);
      }

      return {
        success: true,
        jobId: input.jobId,
        message: 'Job deleted successfully',
      };
    }),

  /**
   * Get job queue stats
   */
  getJobStats: publicProcedure.query(() => {
    return prefectService.getJobStats();
  }),
});
